net/http/src/THttpLongPollEngine.cxx

```diff
@@ -48,12 +57,73 @@ void THttpLongPollEngine::ClearHandle()
    }
 }
 
+//////////////////////////////////////////////////////////////////////////
+/// Create raw buffer which should be send as reply
+/// For the raw mode all information must be send via binary response
+
+void *THttpLongPollEngine::MakeBuffer(const void *buf, int &len, const char *hdr)
+{
+   if (!fRaw) {
+      void *res = malloc(len);
+      memcpy(res, buf, len);
+      return res;
+   }
+
+   int hdrlen = hdr ? strlen(hdr) : 0;
+   std::string hdrstr = "bin:";
+   hdrstr.append(std::to_string(hdrlen));
+
+   while ((hdrstr.length() + 1 + hdrlen) % 8 != 0)
+      hdrstr.append(" ");
+   hdrstr.append(":");
+   if (hdrlen > 0)
+      hdrstr.append(hdr);
+
+   void *res = malloc(hdrstr.length() + len);
+   memcpy(res, hdrstr.c_str(), hdrstr.length());
+   memcpy((char *)res + hdrstr.length(), buf, len);
+   len += hdrstr.length();
+   return res;
+}
+
 //////////////////////////////////////////////////////////////////////////
 /// Send binary data via connection - not supported
 
-void THttpLongPollEngine::Send(const void * /*buf*/, int /*len*/)
+void THttpLongPollEngine::Send(const void *buf, int len)
+{
+   void *buf2 = MakeBuffer(buf, len);
+
+   if (fPoll) {
+      fPoll->SetContentType("application/x-binary");
+      fPoll->SetBinData(buf2, len);
+      fPoll->NotifyCondition();
+      fPoll = nullptr;
+   } else {
+      fQueue.emplace_back(buf2, len);
+      if (fQueue.size() > 100)
+         Error("SendCharStar", "Too many send operations %d, check algorithms", (int)fQueue.size());
+   }
+}
+
+//////////////////////////////////////////////////////////////////////////
+/// Send binary data with text header via connection - not supported
+
+void THttpLongPollEngine::SendHeader(const char *hdr, const void *buf, int len)
 {
-   Error("Send", "Binary send is not supported, use only text");
+   void *buf2 = MakeBuffer(buf, len, hdr);
+
+   if (fPoll) {
+      fPoll->SetContentType("application/x-binary");
+      fPoll->SetBinData(buf2, len);
+      if (!fRaw)
+         fPoll->SetExtraHeader("LongpollHeader", hdr);
+      fPoll->NotifyCondition();
+      fPoll = nullptr;
```

Comment:
I will never get used to this. What happens to `fPoll`, who owns it now? This is really a fairly uncommon coding style. So much so that I would say: please rewrite this.

Reply:
This is not easy.  Here key feature is suspend/resume logic. 

Logic is following - THttpCallArg object is owned by THttpServer and provided for processing to the client code. At any time I can call THttpCallArg::SetPostponed() method to mark object as pending - server will not send any reply for that request. At any moment later I can assign result to that request and call THttpCallArg::NotifyCondition(). Only then THttpServer send a reply to client.

Ownership of THttpCallArg is fully in THttpServer hands. Of course, keeping temporary pointer on THttpCallArg in client code is not safe, but the only alternative is shared_ptr<THttpCallArg>, which means significant change in major THttpServer API.